pkg/server/server.go

```diff
@@ -0,0 +1,184 @@
+package server
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/gorilla/mux"
+	"github.com/xenitab/azdo-proxy/pkg/auth"
+)
+
+type AzdoServer struct {
+	logger logr.Logger
+	port   string
+	proxy  *httputil.ReverseProxy
+	authz  auth.Authorization
+}
+
+func NewAzdoServer(logger logr.Logger, port string, authz auth.Authorization) *AzdoServer {
+	director := func(req *http.Request) {
+		token, err := tokenFromRequest(req)
+		if err != nil {
+			logger.Error(err, "token not present in request")
+			return
+		}
+		target, err := authz.TargetForToken(token)
+		if err != nil {
+			logger.Error(err, "could not get target for token")
+			return
+		}
+
+		targetQuery := target.RawQuery
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+		req.URL.Path, req.URL.RawPath = joinURLPath(target, req.URL)
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+	}
+
+	return &AzdoServer{
+		logger: logger.WithName("azdo-server"),
+		port:   port,
+		proxy:  &httputil.ReverseProxy{Director: director},
+		authz:  authz,
+	}
+}
+
+// ListenAndServe starts the HTTP server on the specified port
+func (a *AzdoServer) ListenAndServe(stopCh <-chan struct{}) {
+	a.logger.Info("Starting git proxy", "port", a.port)
+	router := mux.NewRouter()
+	router.HandleFunc("/readyz", readinessHandler(a.logger)).Methods("GET")
+	router.HandleFunc("/healthz", livenessHandler(a.logger)).Methods("GET")
+	router.PathPrefix("/").HandlerFunc(proxyHandler(a.logger, a.proxy, a.authz))
+	srv := &http.Server{Addr: fmt.Sprintf(":%s", a.port), Handler: router}
+
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			a.logger.Error(err, "azdo-proxy server crashed")
+			os.Exit(1)
+		}
+	}()
+
+	<-stopCh
```

Comment:
If we moved this out to the main function, then we would be able to start more long running functions before and gracefully shut them down.